File: server/lib/reservation.ts
import { buildSchema, GraphQLError } from 'graphql';
import * as lodash from 'lodash';

import logger from './logger';
import { DemoContext, ReservationCreateInput, ReservationStatus, ReservationUpdateInput, User, UserType } from './common';
import { addReservation, listReservations, updateReservation } from './dal';


const typeDefs = buildSchema(`
    scalar Date
    enum ReservationStatus {
        Created
        Completed
        Canceled
    }
    input ReservationCreateInput {
        guestName: String!
        guestPhone: String!
        arrivalTime: Date!
        tableSize: Int!
    }
    input ReservationUpdateInput {
        id: String!
        guestName: String!
        guestPhone: String!
        arrivalTime: Date!
        tableSize: Int!
        status: ReservationStatus!
    }
    type Reservation {
        id: String!
        guestName: String!
        guestPhone: String!
        arrivalTime: Date!
        tableSize: Int!
        status: ReservationStatus!
    }
    type Query {
        reservations(begin: Date, end: Date): [Reservation]!
    }
    type Mutation {
        addReservation(input: ReservationCreateInput!): Reservation
        updateReservation(input: ReservationUpdateInput): Reservation
    }
`);
const resolvers = {
    Query: {
        reservations: async (_root: any, args: any, context: DemoContext) => {
            if (!context.user) {
                logger.error('addReservation: user not found in context.')
                throw new GraphQLError(
                    'user not found in context.',
                    { extensions: { code: 'Bad Request', http: { status: 400 } } }
                );
            }
            const selector: any = {};
            const user = context.user as User;
            if (user.userType === UserType.Guest) {
                lodash.set(selector, 'guestName.$eq', user.name);
            }
            if (args.begin) {
                lodash.set(selector, 'arrivalTime.$gt', (args.begin as Date).getTime());
            }
            if (args.end) {
                lodash.set(selector, 'arrivalTime.$lt', (args.end as Date).getTime());
            }
            lodash.set(selector, 'status.$in', [ReservationStatus.Created, ReservationStatus.Completed]);
            logger.debug('reservations: selector(%j)', selector);
            const retval = await listReservations(selector);
            logger.debug('reservations: done with result(%j)', retval);
            return retval;
        },
    },
    Mutation: {
        addReservation: async (_root: any, args: any, context: DemoContext) => {
            if (!context.user) {
                logger.error('addReservation: user not found in context.')
                throw new GraphQLError(
                    'user not found in context.',
                    { extensions: { code: 'Bad Request', http: { status: 400 } } }
                );
            }
            const input: ReservationCreateInput = args.input as ReservationCreateInput;
            const user = context.user as User;
            if ((user.userType === UserType.Guest) && (user.name != input.guestName)) {
                logger.error('addReservation: user name is not same with the guest name.')
                throw new GraphQLError(
                    'Wrong guest name',
                    { extensions: { code: 'Bad Request', http: { status: 400 } } }
                );
            }
            logger.debug('addReservation: create input(%j)', input);
            const retval = await addReservation(input);
            logger.debug('addReservation: done with result(%j)', retval);
            return retval;
        },
        updateReservation: async (root: any, args: any, context: DemoContext) => {
            if (!context.user) {
                logger.error('updateReservation: user not found in context.')
                throw new GraphQLError(
                    'user not found in context.',
                    { extensions: { code: 'Bad Request', http: { status: 400 } } }
                );
            }
            const input: ReservationUpdateInput = args.input as ReservationUpdateInput;
            const user = context.user as User;
            if ((user.userType === UserType.Guest) && (user.name != input.guestName)) {
                logger.error('updateReservation: user name is not same with the guest name.')
                throw new GraphQLError(
                    'Wrong guest name',
                    { extensions: { code: 'Bad Request', http: { status: 400 } } }
                );
            }
            logger.debug('updateReservation: create input(%j)', input);
            const retval = await updateReservation(input, user.userType);
            logger.debug('updateReservation: done with result(%j)', retval);
            return retval;
        },
    }
};

export { typeDefs, resolvers }

File: fastapi-server/reservation.py
import strawberry
import typing
import dal
from fastapi import Depends, Request
from strawberry.fastapi import GraphQLRouter
from common import Date, ReservationStatus, Reservation, ReservationCreateInput, ReservationUpdateInput, User, users, UserType
from strawberry.exceptions import StrawberryException


@strawberry.type
class Query:
    @strawberry.field
    async def reservations(
        self,
        info: strawberry.Info,
        begin: Date | None = None,
        end: Date | None = None,
        status: ReservationStatus | None = None,
    ) -> typing.List[Reservation]:
        user: User | None = info.context["user"]
        if user is None:
            raise StrawberryException("UNAUTHENTICATED")
        selector = {}
        if user.userType == UserType.Guest:
            selector["guestName"] = selector.get("guestName", {})
            selector["guestName"].update({"$eq": user.name})
        if begin is not None:
            selector["arrivalTime"] = selector.get("arrivalTime", {})
            selector["arrivalTime"].update({"$gte": begin})
        if end is not None:
            selector["arrivalTime"] = selector.get("arrivalTime", {})
            selector["arrivalTime"].update({"$lte": end})
        if status is not None:
            selector["status"] = selector.get("status", {})
            selector["status"].update({"$eq": status})
        documents = dal.listReservations(selector)
        retval = []
        for document in documents:
            retval.append(Reservation.fromDocument(document))
        return retval


@strawberry.type
class Mutation:
    @strawberry.mutation
    async def addReservation(self, info: strawberry.Info, input: ReservationCreateInput) -> Reservation:
        user: User | None = info.context["user"]
        if user.userType == UserType.Guest:
            if user.name != input.guestName:
                raise StrawberryException("Wrong Name")
        document = input.toInsertDocument()
        document["_id"] = dal.addReservation(document)
        return Reservation.fromDocument(document)

    @strawberry.mutation
    async def updateReservation(input: ReservationUpdateInput) -> Reservation:
        query_filter, update_operation = input.toUpdateFilters()
        dal.updateReservation(query_filter, update_operation)
        return input.toReservation()


reservationSchema = strawberry.Schema(query=Query, mutation=Mutation)


def user_dependency(request: Request) -> User | None:
    results = list(filter(lambda u: u.id == request.headers["authorization"], users))
    if len(results) <= 0:
        return None
    return results[0]


async def get_context(user=Depends(user_dependency)):
    return {"user": user}


reservationApp = GraphQLRouter(reservationSchema, context_getter=get_context)
